### DISCUSSION TITLE
ability to view flow versions and open flow builder from the cli

### DISCUSSION THREAD

**Discussion by GusjkoA** [forcedotcom/cli#2519] [Ideas]
`sf org open --source-file` allows opening Visualforce or Lightning app page. It'd be great to expand this function to support flows. Ideally, an extra flag would allow you to select a specific version, latest or active.

Additionally, it'd be great to have a command to list all flows and versions, or available versions for a given flow,

**Comment by mshanemc** [comment]
builder URLs example using the current version (easy)
/builder_platform_interaction/flowBuilder.app?flowId=3018F000000TOomQAG

[query the tooling API for flows, find the one that matches the filename]

non-current versions are harder.  We'd need extra queries to get the definitions, an extra flag to get let you specify one of he versions, etc.  
/builder_platform_interaction/flowBuilder.app?flowDefId=3008F000000EybH&flowId=3018F000000TOoXQAW&sfdcIFrameOrigin=https%3A%2F%2Fbusiness-enterprise-6863-dev-ed.scratch.lightning.force.com&clc=1

so you'll get the simple version (open the Flow), at least for now.

**Reply by mshanemc** [reply]
https://github.com/salesforcecli/plugin-org/pull/829

**Reply by GusjkoA** [reply]
That's a very helpful addition, thanks!

**Reply by GusjkoA** [reply]
@mshanemc , I've just checked the PR and found the use of an unbound query for Flows. I don't think it will work as running this tooling query results in `MALFORMED_QUERY: When retrieving results with Metadata or FullName fields, the query qualificatio​ns must specify no more than one row for retrieval. Result size:` exception in an org with multiple flows.

Instead, I used the following code in the local custom extension:
```
const flow = await conn.singleRecordQuery<{ DurableId: string }>(
            `SELECT DurableId FROM FlowVersionView WHERE FlowDefinitionView.ApiName = '${path.basename(file, '.flow-meta.xml')}' ORDER BY VersionNumber DESC LIMIT 1`
          );
```

**Reply by mshanemc** [reply]
Good catch, thanks.  How did you get multiple records on the toolingAPI query for Flow?

**Reply by GusjkoA** [reply]
@mshanemc , I used FlowVersionView instead -it's accessible through standard API. DurableId is the flow id, i.e. `/builder_platform_interaction/flowBuilder.app?flowId=${flow.DurableId}` The same query can be used to query for explicitly active version, e.g. `SELECT DurableId, Status FROM FlowVersionView WHERE FlowDefinitionView.ApiName = '${path.basename(file, '.flow-meta.xml')}' ORDER BY Status ASC, VersionNumber DESC LIMIT 1` - where Status ASC will ensure the active flow appears at the top.

**Comment by jshackell-sfdc** [comment]
This issue is fixed in CLI version [2.14.6 (Oct 25, 2023)](https://github.com/forcedotcom/cli/blob/main/releasenotes/README.md#2146-oct-25-2023-stable).
